cpg_pipes/pipeline/pipeline.py

```diff
@@ -0,0 +1,960 @@
+        `id` is used to extract the value when the result is a dictionary.
+        """
+        if id is not None:
+            if not isinstance(self.data, dict):
+                raise ValueError(
+                    f'{self.data} is not a dictionary, can\'t get "{id}".'
+                )
+            return cast(dict, self.data)[id]
+
+        if isinstance(self.data, dict):
+            res = cast(dict, self.data)
+            if len(res.values()) > 1:
+                raise ValueError(
+                    f'{res} is a dictionary with more than 1 element, '
+                    f'please set the `id` parameter'
+                )
+            return list(res.values())[0]
+
+        return self.data
+
+    def as_path(self, id=None) -> Path:
+        """
+        Cast the result to path. Though exception if failed to cast.
+        `id` is used to extract the value when the result is a dictionary.
+        """
+        res = self.as_path_or_resource(id)
+        if not isinstance(res, CloudPath | pathlib.Path):
+            raise ValueError(f'{res} is not a path.')
+        return cast(Path, res)
+
+    def as_resource(self, id=None) -> hb.Resource:
+        """
+        Cast the result to Hail Batch Resource, or throw an error if the cast failed.
+        `id` is used to extract the value when the result is a dictionary.
+        """
+        res = self.as_path_or_resource(id)
+        if not isinstance(res, hb.Resource):
+            raise ValueError(f'{res} is not a Hail Batch Resource.')
+        return cast(hb.Resource, res)
+
+    def as_dict(self) -> dict[str, Path | hb.Resource]:
+        """
+        Cast the result to a dictionary, or throw an error if the cast failed.
+        """
+        if not isinstance(self.data, dict):
+            raise ValueError(f'{self.data} is not a dictionary.')
+        return self.data
+
+    def as_resource_dict(self) -> dict[str, hb.Resource]:
+        """
+        Cast the result to a dictionary of Hail Batch Resources, 
+        or throw an error if the cast failed
+        """
+        return {k: self.as_resource(id=k) for k in self.as_dict()}
+
+    def as_path_dict(self) -> dict[str, Path]:
+        """
+        Cast the result to a dictionary of strings, 
+        or throw an error if the cast failed.
+        """
+        return {k: self.as_path(id=k) for k in self.as_dict()}
+
+
+# noinspection PyShadowingNames
+class StageInput:
+    """
+    Represents an input for a stage run. It wraps the outputs of all required upstream stages
+    for corresponding targets (e.g. all GVCFs from a HaploytypeCallerStage
+    for a JointCallingStage, along with Hail Batch jobs).
+
+    An object of this class is passed to the public `queue_jobs` method of a Stage, 
+    and can be used to query dependency files and jobs.
+    """
+
+    def __init__(self, stage: 'Stage'):
+        self.stage = stage
+        self._results_by_target_by_stage: dict[str, dict[str, StageOutput]] = {}
```

Comment:
This variable name could be clarified with by calling it `_result_by_stage_then_by_target`, which is what the usage implies.

Reply:
Just thinking more about this, why do the results have to be keyed by Stage?

Reply:
For the scenarios when a stage depends on multiple other stages